qutebrowser/browser/commands.py

```diff
@@ -1728,3 +1768,252 @@ def fullscreen(self, leave=False):
+            self._set_current_index(idx)
+
+    @cmdutils.register(instance='command-dispatcher', scope='window')
+    def tree_tab_toggle_hide(self):
+        """If the current tab's children are shown hide them, and vice-versa.
+
+        This toggles the current tab's node's `collapsed` attribute.
+        """
+        tabwidget = self._tabbed_browser.widget
+        tab = self._current_widget()
+        if not tab.node.children:
+            return
+        collapsed = tab.node.collapsed
+
+        if collapsed:
+            self._tree_tab_show(tab)
+        else:
+            self._tree_tab_hide(tab)
+
+        tabwidget.update_tree_tab_positions()
+        tabwidget.update_tab_titles()
+
+    def _tree_tab_show(self, tab):
+        """Shows a tab that was previously collapsed through _tree_tab_hide.
+
+        This puts all the descendants of the tab back at the right index and
+        under the right parent.
+
+        Note: this does NOT update tab positions or titles. You have to do it yourself.
+
+        """
+        tabwidget = self._tabbed_browser.widget
+        cur_idx = self._tabbed_browser.widgets().index(tab)
+        order = notree.TraverseOrder.PRE
+        descendents = list(tab.node.traverse(order, False))[1:]
+        for descendent in descendents:
+            cur_tab = descendent.value
+            cur_parent = descendent.parent
+            name = cur_tab.title()
+            icon = cur_tab.icon()
+            tabwidget.insertTab(cur_idx + 1, cur_tab, icon, name)
+            cur_tab.node.parent = cur_parent  # insertTab resets node
+            cur_idx += 1
+        tab.node.collapsed = False
+
+    def _tree_tab_hide(self, tab):
+        """Collapses a tab, hiding all its children and setting tab.node.collapsed.
+
+        Note: this does NOT update tab positions or titles. You have to do it yourself.
+        """
+        tabwidget = self._tabbed_browser.widget
+        order = notree.TraverseOrder.POST
+        descendents = list(tab.node.traverse(order, False))[:-1]
+        for descendent in descendents:
+            cur_tab = descendent.value
+            idx = self._tabbed_browser.widgets().index(cur_tab)
+            tabwidget.removeTab(idx)
+        tab.node.collapsed = True
+
+    def _tree_tab_cycle_hide(self, node):
+        """Utility function for tree_tab_cycle_hide."""
+        # height = node.height  # height is always rel_height
+        if node.collapsed:
+            self._tree_tab_show(node.value)
+            for d in node.traverse():
+                self._tree_tab_show(d.value)
+            return
+
+        def rel_depth(n):
+            return n.depth - node.depth
+        levels = {}
+        for d in node.traverse():
+            r_depth = rel_depth(d)
+            if r_depth not in levels:
+                levels[r_depth] = []
+            levels[r_depth].append(d)
+        target = 0
+        for level in list(reversed(sorted(levels)))[1:]:
```

Comment:
I think you can simplify this by using `key=` in sorted and splitting both level and nodes in the for loop